[AMDGPU][Scheduler] Scoring system for rematerialization candidates

lucas-rami opened the pull request:
This is a significant refactoring of the scheduler's rematerialization stage meant to improve rematerialization capabilities and lay strong foundations for future improvements.

The core contribution is a scoring system to estimate the benefit of each rematerialization candidate. This score takes into account

1. region frequencies,
2. estimated impact on function latency (based on instruction latency), and
3. estimated impact on RP in regions that have excess RP with respect to an objective.

The overall stage's flow is as follows. After setting its RP objective (to reach a specific occupancy, or to eliminate spilling), the stage collects all rematerialization candidates in the function. Among those candidates it identifies and rematerializes all those that are always beneficial to rematerialize regardless of the stage's objective. Then, it scores all other candidates and iteratively rematerializes them in decreasing score order until it reaches the RP objective in all regions. As with the current implementation, all affected regions are then rescheduled, and rematerializations conditionally rolled back if they end up not benefiting performance.

New tests in `machine-scheduler-rematerialization-scoring.mir` showcase how the scoring system dictates which rematerialization are the most beneficial and therefore performed first.

lucas-rami committed:
Scoring system for rematerializations

lucas-rami committed:
Format

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1089,33 +1102,224 @@ bool ClusteredLowOccStage::initGCNSchedStage() {
+
+void PreRARematStage::RematReg::print(
+    const DenseMap<MachineInstr *, unsigned> &MIRegion) const {
+  REMAT_DEBUG(dbgs() << "  [" << MIRegion.at(DefMI) << "] " << *DefMI);
+  REMAT_DEBUG(dbgs() << "    -> used in [" << UseRegion << "] " << *UseMI);
+  const unsigned NumRegions = Live.size();
+  REMAT_DEBUG(dbgs() << "    Guaranteed RP reduction in:");
+  for (unsigned I = 0; I < NumRegions; ++I) {
+    if (isBeneficialRegion(I))
+      dbgs() << " [" << I << "]";
+  }
+  dbgs() << '\n';
+  REMAT_DEBUG(dbgs() << "    Possible RP reduction in:");
+  for (unsigned I = 0; I < NumRegions; ++I) {
+    if (isMaybeBeneficialRegion(I))
+      dbgs() << " [" << I << "]";
+  }
+  dbgs() << '\n';
```
Ditto, REMAT_DEBUG redundant with the ones at the use points

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1089,33 +1102,224 @@ bool ClusteredLowOccStage::initGCNSchedStage() {
+void PreRARematStage::printTargetRegions(bool PrintAll) const {
+  if (PrintAll) {
+    for (auto [I, Target] : enumerate(RPTargets))
+      REMAT_DEBUG(dbgs() << "  [" << I << "] " << Target << '\n');
```
You only use this under REMAT_DEBUG uses, so can drop all the REMAT_DEBUGs here

arsenm replied:
Then wrap those dangling uses in the _DEBUG blocks, they don't need their own ifndef NDEBUGs

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1782,30 +1950,34 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
+      // either rematerializing the candidates in careful ordering, or
+      // deferring the MBB RP walk until the entire chain has been
+      // rematerialized.
+      MachineOperand &UseFirstMO = UseMI->getOperand(0);
+      if ((UseFirstMO.isReg() && RematRegSet.contains(UseFirstMO.getReg())) ||
+          llvm::any_of(DefMI.operands(), [&RematRegSet](MachineOperand &MO) {
+            return MO.isReg() && RematRegSet.contains(MO.getReg());
```
This looks backwards? The UseMI is looking at the assumed single def, and DefMI is looking at all of its operands? Should one or both of these checks be verifying it's a use/def?

lucas-rami replied:
I think the implementation is right (though it is confusing, so I refactored it to make it clearer).

```cpp
UseFirstMO.isReg() && RematRegSet.contains(UseFirstMO.getReg())
```

This ensures that `UseMI` is not itself an instruction we have already added to the rematerializable list. If the first operand is not a def reg then the instruction couldn't have been considered rematerializable, so `RematRegSet` won't contain the register.

```cpp
llvm::any_of(DefMI.operands(), [&RematRegSet](MachineOperand &MO) {
  return MO.isReg() && RematRegSet.contains(MO.getReg())
})
```

This ensures that no operand of `DefMI` is defined by an instruction we have added to the rematerializable list. We know that the first operand will clear the test since we only consider registers with a single def as potentially rematerializable, so the defined register is guaranteed to not be in `RematRegSet` already.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1817,106 +1989,146 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
+
+  // Store the register's lane bitmask.
+  unsigned SubReg = DefMI->getOperand(0).getSubReg();
+  Mask = SubReg ? DAG.TRI->getSubRegIndexLaneMask(SubReg)
```
Is using LaneBitmask::getAll good enough here?

lucas-rami replied:
I can in fact simplify this whole thing to `getSubRegIndexLaneMask(SubReg)` since it can handle the 0 case, thanks for the catch.

lucas-rami replied:
This actually doesn't work since the lane mask returned when subreg is 0 is ~0U, yielding a very large/incorrect number when calling `getNumCoveredRegs` on it. By sheer luck this did not break tests on a07fef2 but only on 8c142c8 which changes rematerialization priorities in some tests when multiple rematerialization candidates have the same score. I rollbacked the change in 70cd6f6.

arsenm replied:
getNumCoveredRegs is a broken API. You're not allowed to rely on lanemask values this way. You're only allowed to check for overlaps. A LaneMask on its own doesn't mean anything, you need to know the associated register class.

We probably should have a tablegenerated version of this function, or find a different way to get a register count at the use point

lucas-rami replied:
Thanks for the context, this is only used for the scoring heuristic and for the sake of time I just resorted to using the register class to get the register size, without accounting for a potential subregister. Left a FIXME for future reference.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1817,106 +1989,146 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
+  const SIInstrInfo *TII = MF.getSubtarget<GCNSubtarget>().getInstrInfo();
+  MachineInstr &DefMI = *Remat.DefMI;
+  Register Reg = DefMI.getOperand(0).getReg();
+  const TargetRegisterClass *RC = DAG.MRI.getRegClass(Reg);
+  Register NewReg = DAG.MRI.createVirtualRegister(RC);
```
cloneVirtualRegister? Could this lose WWM flags?

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1940,65 +2152,80 @@ void PreRARematStage::rematerialize() {
+  unsigned DefRegion = MIRegion.at(Remat->DefMI);
+  MachineBasicBlock *MBB = RegionBB[DefRegion];
+  Register Reg = RematMI->getOperand(0).getReg();
+  const TargetRegisterClass *RC = DAG.MRI.getRegClass(Reg);
+  Register NewReg = DAG.MRI.createVirtualRegister(RC);
```
Should this be cloneVirtualRegister?

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1940,65 +2152,80 @@ void PreRARematStage::rematerialize() {
+  // region, but it should not matter much.
+  MachineBasicBlock::iterator InsertPos(DAG.Regions[DefRegion].second);
+  TII->reMaterialize(*MBB, InsertPos, NewReg, 0, *RematMI, *DAG.TRI);
+  REMAT_DEBUG(dbgs() << "[" << DefRegion << "] Re-rematerialized as "
```
```suggestion
  REMAT_DEBUG(dbgs() << '[' << DefRegion << "] Re-rematerialized as "
```

arsenm commented on llvm/test/CodeGen/AMDGPU/machine-scheduler-rematerialization-scoring.mir:
```diff
@@ -0,0 +1,535 @@
+#           +---+
+
+--- |
+  define void @favor_always_benef() {
+    ret void
+  }
+  define void @favor_live_through_in_high_freq_region() {
+    ret void
+  }
+  define void @use_only_region_possible() {
+    ret void
+  }
```
Don't need IR section

arsenm commented on llvm/test/CodeGen/AMDGPU/machine-scheduler-rematerialization-scoring.mir:
```diff
@@ -0,0 +1,535 @@
+    liveins: $vgpr0, $sgpr0_sgpr1
+
+    %mem_addr:sgpr_64(p4) = COPY $sgpr0_sgpr1
+    %loop_if_bound:vgpr_32(s32) = COPY $vgpr0
```
Drop the accidental LLTs on all the registers

lucas-rami committed:
Address feedback + fix failing test

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -28,10 +28,21 @@
+#include "llvm/MC/TargetRegistry.h"
 #include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/raw_ostream.h"
+#include <deque>
```
Nowhere, thanks for the catch.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1089,33 +1100,224 @@ bool ClusteredLowOccStage::initGCNSchedStage() {
+    // (Re-)Score and (re-)sort all remats in increasing score order.
+    for (ScoredRemat &Remat : ScoredRemats)
+      Remat.update(TargetRegions, RPTargets, RegionFreq, !TargetOcc);
+    stable_sort(ScoredRemats);
```
is stable_sort really necessary, it's better to add tie breakers to the sort function

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1691,87 +1896,51 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+  // remat chains checks.
+  DenseSet<Register> RematRegSet;
+  auto IsMORematable = [&RematRegSet](const MachineOperand &MO) -> bool {
+    return MO.isReg() && RematRegSet.contains(MO.getReg());
```
Should this also only consider isUse or readsReg? Or otherwise ignore undef uses?

lucas-rami replied:
Any of `DefMI`'s register operands being in RematRegSet implies isUse since rematerializable registers are right now limited to having a single defining instruction that, in the subreg case, doesn't read other parts of the register. `UseMI`'s first register operand being in RematRegSet implies isDef. So I think these checks would be redundant.

I don't think we can currently allow undef uses of a rematable register, otherwise we will have liverange issues. Some of these limitations will go away once we support remat chains, which I am currently working on.

lucas-rami committed:
Remove REMAT_DEBUG and break ties in score

lucas-rami committed:
Fix failing tests and rollback mask change

lucas-rami commented:
ping

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.h:
```diff
@@ -432,66 +436,208 @@ class ClusteredLowOccStage : public GCNSchedStage {
+    /// Latency gain induced by rematerializing the instruction. Does not
+    /// include estimated spilling cost of *not* rematerializing (save/restore
+    /// to/from stack).
+    std::optional<int> InstrLatencyGain = std::nullopt;
```
```suggestion
    std::optional<int> InstrLatencyGain;
```

Can you avoid optional, and use -1 or 0 to naturally get the same effect without special checks?

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1691,87 +1896,51 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
-  DenseSet<unsigned> RematRegs;
+  // Set of registers already marked for potential remterialization; used for
+  // remat chains checks.
+  DenseSet<Register> RematRegSet;
```
```suggestion
  SmallSet<Register, 4> RematRegSet;
```

arsenm replied:
Maybe should be tracking this on RegSubRegPair. Eventually we really need to improve rematerialize for sub registers

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1817,106 +1987,145 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
+                                          const GCNSubtarget &ST,
+                                          const TargetInstrInfo &TII)
+    : Remat(Remat) {
+  const InstrItineraryData *Itin = ST.getInstrItineraryData();
```
Isn't this the deprecated one? Should this be using SchedModel instead?

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1089,33 +1100,224 @@ bool ClusteredLowOccStage::initGCNSchedStage() {
+      REMAT_DEBUG(dbgs() << "[" << MIRegion[Remat.DefMI] << "] REMAT *" << Score
+                         << "* | " << *Remat.DefMI);
+      MachineInstr *RematMI = rematerialize(Remat, RecomputeRP);
+      // Every rematerialization done with the objective of increasing occupancy
+      // increases latency. If we don't manage to increase occupancy, we want to
```
but it is true when moving into loops or if it breaks coalescing

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1817,106 +1987,145 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
+  Register Reg = Remat->DefMI->getOperand(0).getReg();
+  unsigned RPScore = 0;
+  for (unsigned I : TargetRegions.set_bits()) {
+    unsigned Freq = std::max(RegionFreq[I], static_cast<uint64_t>(1));
```
Checking the frequency of the affected region only makes sense if we're going to reduce spilling. 

For the occupancy case, this info is basically irrelevant.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1817,106 +1987,145 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
-      GCNRPTracker::LiveRegSet &RegionLiveIns = DAG.LiveIns[I];
+  // The estimated RP reduction is directly proportional to the size of the
+  // rematerializable register.
+  setRPScore(RPScore * SIRegisterInfo::getNumCoveredRegs(Remat->Mask));
```
For both the occupancy and spilling cases, I would think we would want to prioritize remats into less hot blocks.

lucas-rami replied:
> For the occupancy case, the frequency of the intermediate target blocks doesn't matter -- we remove latency from the original def region, and we add latency to the remat region. We shouldn't care about the intermediate frequency as doing a remat has no impact on those regions -- if we do enough remats we will increase occupancy which is function wide.
> For the spilling case, the frequencey of the intermediate target blocks does matter. If our RP is above the threshold, then we will spill somewhere, we would prefer not to spill in the hottest blocks. Thus, we should prioritize reducing RP in the highest frequency regions first.

Thanks a lot for the explanation, that makes sense to me now. Updated the scoring method to reflect this.

> IMO, the scoring model should take into account the (perhaps in order of priority):
> 1. maximal freuency of affected regions (if spilling)
> 2. delta between def / remat frequency
> 3. the number of target regions affected
> 4. reg / lanemask width.

With the above remark in mind those make more sense than what I had indeed. Implemented exactly those scoring components with those priorities (with merged 3/4).

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1817,106 +1987,145 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
+    // most cases cheaper than spilling. We still give a bonus to remats for
+    // which we are able to do the calculation.
+    if (InstrLatencyGain && *InstrLatencyGain < 0) {
+      int SpillLatencyGain = SaveCost * Remat->DefFrequency;
```
This spill cost model seems a bit flaky -- can we get away with just checking the def vs remat frequency and the latency?

lucas-rami replied:
I have moved the calculation to `ScoredRemat::getLatencyGain` but it is the same idea i.e. I'm comparing the latency difference induced by the rematerializable MI being moved to a different region to the additional latency that would come from saving/restoring to/from stack. The former should be smaller than the latter in almost all cases. Does that make sense?

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1782,31 +1951,32 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
 
-      // We only support rematerializing virtual registers with one definition.
+      // We only support rematerializing virtual registers with one
+      // definition.
```
Not directly related to this PR, but I think we also need to check either that the use operands of the remat candidate only have 1 def, or, preferably, when doing the remat we check that the reaching def of the uses is the same as the original def.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1151,15 +1352,19 @@ void UnclusteredHighRPStage::finalizeGCNSchedStage() {
 
-  // Skip empty scheduling regions (0 or 1 schedulable instructions).
-  if (DAG.begin() == DAG.end() || DAG.begin() == std::prev(DAG.end()))
+  // Skip regions with 1 schedulable instruction.
+  if (DAG.begin() == std::prev(DAG.end()))
     return false;
```
Unfortunately no, `ScheduleDAGMILive::enterRegion` expects to be called even for single-instruction regions to maintain some internal state.

lucas-rami committed:
Address more feedback

- Use SchedModel instead of instruction itinerary
- Don't use getNumConvertedRegs to get number of regs, use RC instead.
- Clarify some comments.
- Other minor changes.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.h:
```diff
@@ -432,66 +436,208 @@ class ClusteredLowOccStage : public GCNSchedStage {
+    }
+
+    /// Returns whether is is always beneficial to rematerialize this register.
+    bool isAlwaysBeneficial() const {
```
All registers we consider rematerializable have a single use at this point, added a comment on `PreRARematStage` to clarify all of the current constraints on rematable regs. My next PR will add support for remat with multiple uses, and will take into account what you are pointing to here.

jrbyrnes replied:
Sure -- I was just suggesting to future proof this method against the eventual expansion into multi-user remat. But, general awareness should suffice.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.h:
```diff
@@ -432,66 +436,208 @@ class ClusteredLowOccStage : public GCNSchedStage {
+    /// Estimated save/restore latency costs for spilling a register to stack.
+    /// FIXME: These numbers are very arbitrary. Need a good rationale for them,
+    /// which I don't know where to get from.
+    static constexpr int SaveCost = 100, RestoreCost = 100;
```
I am now using the SchedModel for instruction latencies so I also tried using it with `SIInstrInfo::getVectorRegSpill[Save|Restore]Opcode` to get the opcodes of the spilling instructions that would be used to spill any given register. Unfortuantely these opcodes don't have an associated scheduling model so it doesn't work. I ended up using the `MCSchedModel::DefaultHighLatency` constant to estimate spill save/restore cost.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.h:
```diff
@@ -432,66 +436,208 @@ class ClusteredLowOccStage : public GCNSchedStage {
+    }
+
+    /// Returns whether is is always beneficial to rematerialize this register.
+    bool isAlwaysBeneficial() const {
```
Elaborated the function comment to explain why we don't check RP in it.

lucas-rami committed:
Rebase for new test + improve comment

lucas-rami committed:
Refactor scoring system + Remove always benef/latency calc

- Change components of scoring system.
- Remove rematerialization of always beneficial registers.
- Remove latency calculation in scoring.
- Other small changes.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1089,33 +1100,224 @@ bool ClusteredLowOccStage::initGCNSchedStage() {
   }
 
-  if (!canIncreaseOccupancyOrReduceSpill())
+  setObjective();
```
I removed the feature from this PR, making this contribution a little shorter/simpler which is good. I still think this is a useful feature to have but I get your point that it doesn't really belong in the scheduler's remat phase. I also don't know why MachineSink is not taking care of those, it's something I want to look into in the future.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1883,65 +2118,79 @@ void PreRARematStage::rematerialize() {
+    // This save is exact in beneficial regions but optimistic in all other
+    // regions where the register is live.
+    RPTargets[I].saveReg(Reg, Remat.Mask, DAG.MRI);
+    DAG.LiveIns[I].erase(Reg);
```
I think this isn't needed because of the one-def check indeed.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1883,65 +2118,79 @@ void PreRARematStage::rematerialize() {
+    // regions where the register is live.
+    RPTargets[I].saveReg(Reg, Remat.Mask, DAG.MRI);
+    DAG.LiveIns[I].erase(Reg);
+    DAG.RegionLiveOuts.getLiveRegsForRegionIdx(I).erase(Reg);
```
Ditto as well

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1760,106 +1933,168 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
+  return divideCeil(DAG.TRI->getRegSizeInBits(RC), 32);
+}
+
+unsigned PreRARematStage::ScoredRemat::getLatencyGain(
```
Makes sense, removed it.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -2040,6 +2257,13 @@ void GCNScheduleDAGMILive::updateRegionBoundaries(
 }
 
+void GCNScheduleDAGMILive::deleteMI(unsigned RegionIdx, MachineInstr *MI) {
+  updateRegionBoundaries(Regions[RegionIdx], MI, nullptr);
```
I ended up deleting `updateRegionBoundaries` because I was the only user and it was a somewhat long function that can be replaced with a couple simple lines at both use points. However I don't think I changed the implementation by doing that, and I am not seeing an issue when deleting the only instruction from the region in `deleteMI`.

If `MI` is the only instruction in the region we have `Region.second == std::next(MachineBasicBlock::iterator(MI))` so after the function we will have `Region.first == Region.second`. `test_rollback_remats_emptydefregion` in `machine-scheduler-sink-trivial-remats.mir` tests that we can completely empty a region (bb.1) and then rollback instructions to it again and it seems to work without issue.

lucas-rami committed:
Fix tests

lucas-rami committed:
Improve code for frequency-based calculations

- Add some explanatory comments for the calculations.
- Normalize frequencies to minimum to avoid overflows.
- Create helper struct to hold all frequency information (helpful for
  further improvements).

lucas-rami committed:
Add empty region test

lucas-rami committed:
Correctly derive (sub)reg size and frequency fix

In cases where the (normalized) maximum region frequency is lower than
the number of different frequency scores representable (2^16 currently),
the current method of computing the scaling factor shrinks the actually
achievable range of frequency scores available (up to a single
achievable score in the most extreme cases). This is due to integer
rounding when dividing frequencies to fit within the representable
range.

When the maximum frequency is smaller than the maximum score achievable,
use a different rescaling calculation to avoid the expressivity loss.

lucas-rami committed:
Walk over all regions to compute AchievedOcc

lucas-rami committed:
Use std::max for freq calc.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1090,33 +1101,223 @@ bool ClusteredLowOccStage::initGCNSchedStage() {
+      dbgs() << "reduce spilling (minimum target occupancy is "
+             << MFI.getMinWavesPerEU() << ")\n";
+    }
+    printTargetRegions(/*PrintAll=*/TargetRegions.none());
```
I think `TargetRegions.none()` will always be false here -- since `setObjective` would have returned false above?

lucas-rami replied:
Indeed, this also allows me to remove the function argument altogether, thanks.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1090,33 +1101,223 @@ bool ClusteredLowOccStage::initGCNSchedStage() {
+
+    // Peel off registers we already rematerialized from the vector's tail.
+    ScoredRemats.truncate(RematIdx + 1);
+  } while ((updateAndVerifyRPTargets(RecomputeRP) || TargetRegions.any()) &&
```
Is it possible to hit infinite loop ? If we have a `ScoredRemat` that is no longer beneficial, but non-empty `TargetRegions`?

lucas-rami committed:
Simplify score calculation and improve debug

- Removed some debug-only functions from the header and put them as
  close as possible to where they are needed.
- Split bitpacked score into separate components compared 1-to-1.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1760,106 +1931,203 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
+  // The difference between defining and using frequency is in the range
+  // [-MaxDiff, MaxDiff], shift it to [0,2 x MaxDiff] to stay in the positive
+  // range, then rescale to the representable range in the final score.
+  const uint64_t FreqDiff = (MaxDiff + (DefOrOne - UseOrMax));
```
I removed the bitpacking algorithm and simply used three class members. It does simplify the logic a bunch at the cost of extra data and extra comparison/sorting time.

It may not matter much at this point since we typically don't have a lot of rematerialization candidates due to the exsiting limitations but with support for remat with multiple uses and chains the number of candidates is much larger, so these things may become more important more later, and I can always re-introduce that then.

The point of the now removed calculation around `FreqDiff` was to rescale the range of possible differences between defining and using region to the range of representable values in the bitpacked score (as a function of the number of bits allocated to that scoring component). It mapped the largest possible penalty to latency (DefFreq < UseFreq) to 0 and the smallest possible penalty to latency (DefFreq > UseFreq) to 0xFF..FF, with a linear (up to integer rounding) interpolation of the values in between. `favor_lower_frequency` in `machine-scheduler-rematerialization-scoring.mir` shows that we will prefer to rematerialize registers into regions of lower frequency, all other things being equal.

lucas-rami committed:
Early exit on maxWavesPerEU and rebase

lucas-rami committed:
Fix tests and remove arg from TII.rematerialize

nhaehnle commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.h:
```diff
@@ -451,65 +447,215 @@ class ClusteredLowOccStage : public GCNSchedStage {
-  /// PreRARematStage::TargetOccupancy.
-  bool canIncreaseOccupancyOrReduceSpill();
+  /// List of rematerializable registers.
+  SmallVector<RematReg, 16> RematRegs;
```
Prefer not to specify SmallVector sizes explicitly -- unless you have a really good reason for it?

nhaehnle commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.h:
```diff
@@ -451,65 +447,215 @@ class ClusteredLowOccStage : public GCNSchedStage {
+      if (RegionImpact != O.RegionImpact)
+        return RegionImpact < O.RegionImpact;
+      // Break ties using pointer to rematerializable register.
+      return Remat > O.Remat;
```
Breaking ties using pointer values is never acceptable because it causes the compiler to become non-deterministic.

A reasonable alternative would be to just return false for scores that compare equal and use `stable_sort`.

nhaehnle reviewed (commented):
The overall flow of the change looks good to me, but I do have some comments.

So the "rollback" doesn't actually roll back the change reliably. That feels like a problem. I understand it's a little tricky to restore the original instruction in exactly its original place after having deleted it first. Perhaps you can leave the original instruction in place in a modified form that doesn't affect the register pressure estimation?

Should the score be affected by the register size? (This is probably mostly relevant for rematerializing SGPRs. There's no difference in instruction cost between a 32-bit and a 128-bit register, but obviously rematerializing the larger register is likely to have a bigger benefit. But we do have larger VALU instructions as well...)

lucas-rami committed:
Address feedback

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1269,33 +1280,225 @@ bool ClusteredLowOccStage::initGCNSchedStage() {
     return false;
+  if (DAG.MinOccupancy >= MFI.getMaxWavesPerEU() &&
+      !MF.getFunction().hasFnAttribute("amdgpu-num-sgpr") &&
+      !MF.getFunction().hasFnAttribute("amdgpu-num-vgpr"))
```
Deleted the check entirely. The same behavior is achieved by letting `setObjective` determine there is nothing to do on its own and using the subtarget's methods directly.

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1960,106 +2133,176 @@ bool PreRARematStage::canIncreaseOccupancyOrReduceSpill() {
+  // Store the register's lane bitmask.
+  unsigned SubIdx = DefMI->getOperand(0).getSubReg();
+  Mask = SubIdx ? DAG.TRI->getSubRegIndexLaneMask(SubIdx)
+                : DAG.MRI.getMaxLaneMaskForVReg(Reg);
```
I just ended up querying the live-outs map for the register's defining region, since it always contains the mask of interest.

lucas-rami committed:
Rollback without removing MI
